File: hooks/post-tool-use.py
#!/usr/bin/env python3
"""PostToolUse hook - tracks edited files for memory updates.

Fires after Edit, Write, or Bash tool execution. Appends changed file
paths to .claude/mark42/dirty-files for batch processing at turn end.
Produces NO OUTPUT to maintain zero token cost (critical for performance).

Supports configurable trigger modes:
- default: Track Edit/Write/Bash operations (current behavior)
- gitmode: Only track git commits
"""

from __future__ import annotations

import json
import os
import shlex
import subprocess
import sys
from pathlib import Path


def load_config(project_dir: str) -> dict:
    """Load plugin configuration from .claude/mark42/config.json."""
    config_file = Path(project_dir) / ".claude" / "mark42" / "config.json"
    if config_file.exists():
        try:
            with open(config_file) as f:
                return json.load(f)
        except (json.JSONDecodeError, OSError):
            pass
    return {"triggerMode": "default"}


def handle_git_commit(project_dir: str) -> tuple[list[str], dict | None]:
    """Extract context from a git commit.

    Returns: (files, commit_context) where commit_context is {"hash": ..., "message": ...}
    """
    result = subprocess.run(
        ["git", "log", "-1", "--format=%h %s"],
        capture_output=True,
        text=True,
        cwd=project_dir,
    )
    if result.returncode != 0:
        return [], None

    parts = result.stdout.strip().split(" ", 1)
    commit_hash = parts[0]
    commit_message = parts[1] if len(parts) > 1 else ""

    result = subprocess.run(
        ["git", "diff-tree", "--no-commit-id", "--name-only", "-r", "HEAD"],
        capture_output=True,
        text=True,
        cwd=project_dir,
    )
    if result.returncode != 0:
        return [], {"hash": commit_hash, "message": commit_message}

    files = [f.strip() for f in result.stdout.strip().split("\n") if f.strip()]
    files = [str((Path(project_dir) / f).resolve()) for f in files]

    return files, {"hash": commit_hash, "message": commit_message}


def should_track(file_path: str, project_dir: str) -> bool:
    """Check if file should be tracked for memory updates."""
    path = Path(file_path)

    try:
        relative = path.relative_to(project_dir)
    except ValueError:
        return False

    # Exclude .claude/ directory (plugin state files)
    if relative.parts and relative.parts[0] == ".claude":
        return False

    # Exclude CLAUDE.md files anywhere
    if path.name == "CLAUDE.md":
        return False

    return True


def extract_files_from_bash(command: str, project_dir: str) -> list[str]:
    """Extract file paths from Bash commands that modify files."""
    if not command:
        return []

    command = command.strip()

    skip_prefixes = (
        "ls", "cat", "echo", "grep", "find", "head", "tail", "less", "more",
        "cd", "pwd", "which", "whereis", "type", "file", "stat", "wc",
        "git status", "git log", "git diff", "git show", "git branch",
        "git fetch", "git pull", "git push", "git clone", "git checkout",
        "git stash", "git remote", "git tag", "git rev-parse",
        "npm ", "yarn ", "pnpm ", "node ", "python", "pip ", "uv ",
        "cargo ", "go ", "make", "cmake", "docker ", "kubectl ",
        "curl ", "wget ", "ssh ", "scp ", "rsync ",
    )
    if command.startswith(skip_prefixes):
        return []

    shell_operators = ("&&", "||", ";", "|", ">", ">>", "<", "2>", "2>&1")
    files = []

    try:
        tokens = shlex.split(command)
        if not tokens:
            return []

        cmd = tokens[0]

        if cmd == "rm":
            for token in tokens[1:]:
                if token in shell_operators:
                    break
                if not token.startswith("-"):
                    files.append(token)

        elif cmd == "git" and len(tokens) > 1 and tokens[1] == "rm":
            for token in tokens[2:]:
                if token in shell_operators:
                    break
                if not token.startswith("-"):
                    files.append(token)

        elif cmd == "mv" and len(tokens) >= 3:
            for token in tokens[1:]:
                if token in shell_operators:
                    break
                if not token.startswith("-"):
                    files.append(token)
                    break

        elif cmd == "git" and len(tokens) > 2 and tokens[1] == "mv":
            for token in tokens[2:]:
                if token in shell_operators:
                    break
                if not token.startswith("-"):
                    files.append(token)
                    break

        elif cmd == "unlink" and len(tokens) > 1:
            if tokens[1] not in shell_operators:
                files.append(tokens[1])

    except ValueError:
        return []

    resolved = []
    for f in files:
        path = Path(f)
        if not path.is_absolute():
            path = Path(project_dir) / path
        resolved.append(str(path.resolve()))

    return resolved


def main():
    project_dir = os.environ.get("CLAUDE_PROJECT_DIR", "")
    if not project_dir:
        return

    try:
        stdin_data = sys.stdin.read()
        tool_input = json.loads(stdin_data) if stdin_data else {}
    except json.JSONDecodeError:
        tool_input = {}

    tool_name = tool_input.get("tool_name", "")
    tool_input_data = tool_input.get("tool_input", {})

    config = load_config(project_dir)
    trigger_mode = config.get("triggerMode", "default")

    is_git_commit = False
    command = ""
    if tool_name == "Bash":
        command = tool_input_data.get("command", "").strip()
        is_git_commit = "git commit" in command

    if trigger_mode == "gitmode" and not is_git_commit:
        return

    files_to_track = []
    commit_context = None

    if is_git_commit:
        files, commit_context = handle_git_commit(project_dir)
        files_to_track.extend(files)
    elif tool_name in ("Edit", "Write"):
        file_path = tool_input_data.get("file_path", "")
        if file_path:
            files_to_track.append(file_path)
    elif tool_name == "Bash":
        files_to_track = extract_files_from_bash(command, project_dir)
    elif not tool_name:
        file_path = tool_input_data.get("file_path", "")
        if file_path:
            files_to_track.append(file_path)

    if not files_to_track:
        return

    trackable = [f for f in files_to_track if should_track(f, project_dir)]
    if not trackable:
        return

    dirty_file = Path(project_dir) / ".claude" / "mark42" / "dirty-files"
    dirty_file.parent.mkdir(parents=True, exist_ok=True)

    existing: dict[str, str] = {}
    if dirty_file.exists():
        with open(dirty_file) as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                path = line.split(" [")[0] if " [" in line else line
                existing[path] = line

    for file_path in trackable:
        if commit_context:
            ctx = f"[{commit_context['hash']}: {commit_context['message']}]"
            existing[file_path] = f"{file_path} {ctx}"
        elif file_path not in existing:
            existing[file_path] = file_path

    with open(dirty_file, "w") as f:
        for line in existing.values():
            f.write(line + "\n")

    # Write session event (JSON Lines) for capture at session end
    event_file = Path(project_dir) / ".claude" / "mark42" / "session-events"
    event_file.parent.mkdir(parents=True, exist_ok=True)
    import datetime
    event = {"toolName": tool_name, "timestamp": datetime.datetime.now(datetime.timezone.utc).isoformat()}
    if tool_name in ("Edit", "Write") and trackable:
        event["filePath"] = trackable[0]
    elif tool_name == "Bash" and command:
        event["command"] = command[:200]
    with open(event_file, "a") as f:
        f.write(json.dumps(event) + "\n")

    # NO output - zero token cost


if __name__ == "__main__":
    main()

File: internal/adapter/claude/post_tool_use.go
package claude

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/mfenderov/mark42/internal/storage"
)

// PostToolUse runs the PostToolUse hook: track file modifications + session events.
func PostToolUse(projectDir string, input HookInput, opts ...Option) {
	cfg := loadPluginConfig(projectDir)

	hookCfg := &Config{}
	for _, o := range opts {
		o(hookCfg)
	}

	command, isGitCommit := bashCommand(input)
	if cfg.TriggerMode == "gitmode" && !isGitCommit {
		return
	}

	trackable := filterTrackable(filesFromInput(input, command, isGitCommit, projectDir), projectDir)

	m42 := mark42Dir(projectDir)
	_ = os.MkdirAll(m42, 0o755)

	recordSessionEvent(projectDir, input, command, trackable, hookCfg)
	updateDirtyFiles(m42, trackable)

	// CRITICAL: zero stdout output
}

// bashCommand extracts the trimmed command and whether it is a git commit.
func bashCommand(input HookInput) (command string, isGitCommit bool) {
	if input.ToolName != "Bash" {
		return "", false
	}
	cmd, ok := input.ToolInput["command"].(string)
	if !ok {
		return "", false
	}
	command = strings.TrimSpace(cmd)
	return command, strings.Contains(command, "git commit")
}

// filesFromInput determines which files a tool invocation touched.
func filesFromInput(input HookInput, command string, isGitCommit bool, projectDir string) []string {
	switch {
	case isGitCommit:
		// Git commit file extraction requires subprocess — skip for now,
		// just track the event
		return nil
	case input.ToolName == "Bash":
		return extractFilesFromBash(command, projectDir)
	default:
		return filePathArg(input)
	}
}

func filePathArg(input HookInput) []string {
	if fp, ok := input.ToolInput["file_path"].(string); ok && fp != "" {
		return []string{fp}
	}
	return nil
}

func filterTrackable(files []string, projectDir string) []string {
	var trackable []string
	for _, f := range files {
		if shouldTrack(f, projectDir) {
			trackable = append(trackable, f)
		}
	}
	return trackable
}

// recordSessionEvent writes the session event to SQLite when a current session exists.
func recordSessionEvent(projectDir string, input HookInput, command string, trackable []string, hookCfg *Config) {
	sessionName := readCurrentSession(projectDir)
	if sessionName == "" {
		return
	}

	se := storage.SessionEvent{
		ToolName:  input.ToolName,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}
	switch {
	case (input.ToolName == "Edit" || input.ToolName == "Write") && len(trackable) > 0:
		se.FilePath = trackable[0]
	case input.ToolName == "Bash" && command != "":
		se.Command = truncateCommand(command, 200)
	}

	s, owned := getOrUseStore(hookCfg)
	if s == nil {
		return
	}
	if owned {
		defer s.Close()
	}
	_ = s.CaptureSessionEvent(sessionName, se)
}

// truncateCommand cuts a command to n bytes (plain cut, no ellipsis).
func truncateCommand(s string, n int) string {
	if len(s) > n {
		return s[:n]
	}
	return s
}

// updateDirtyFiles merges newly modified files into the dirty-files list.
func updateDirtyFiles(m42 string, trackable []string) {
	if len(trackable) == 0 {
		return
	}

	dirtyPath := filepath.Join(m42, "dirty-files")
	existing := make(map[string]string)
	for _, line := range readLines(dirtyPath) {
		path := line
		if before, _, found := strings.Cut(line, " ["); found {
			path = before
		}
		existing[path] = line
	}

	for _, fp := range trackable {
		if _, ok := existing[fp]; !ok {
			existing[fp] = fp
		}
	}

	var sb strings.Builder
	for _, line := range existing {
		sb.WriteString(line)
		sb.WriteByte('\n')
	}
	_ = os.WriteFile(dirtyPath, []byte(sb.String()), 0o644)
}

func loadPluginConfig(projectDir string) PluginConfig {
	data, err := os.ReadFile(filepath.Join(mark42Dir(projectDir), "config.json"))
	if err != nil {
		return PluginConfig{TriggerMode: "default"}
	}
	var cfg PluginConfig
	if err := json.Unmarshal(data, &cfg); err != nil {
		return PluginConfig{TriggerMode: "default"}
	}
	if cfg.TriggerMode == "" {
		cfg.TriggerMode = "default"
	}
	return cfg
}

func shouldTrack(filePath, projectDir string) bool {
	if !strings.HasPrefix(filePath, projectDir) {
		return false
	}

	rel, err := filepath.Rel(projectDir, filePath)
	if err != nil {
		return false
	}

	parts := strings.SplitN(rel, string(filepath.Separator), 2)
	if len(parts) > 0 && parts[0] == ".claude" {
		return false
	}

	if filepath.Base(filePath) == "CLAUDE.md" {
		return false
	}

	return true
}

func extractFilesFromBash(command, projectDir string) []string {
	command = strings.TrimSpace(command)
	if command == "" || isReadOnlyCommand(command) {
		return nil
	}

	tokens := shellTokenize(command)
	if len(tokens) == 0 {
		return nil
	}

	var files []string
	switch cmd := tokens[0]; {
	case cmd == "rm":
		files = fileArgs(tokens[1:])
	case cmd == "git" && len(tokens) > 1 && tokens[1] == "rm":
		files = fileArgs(tokens[2:])
	case cmd == "mv" && len(tokens) >= 3:
		files = firstFileArg(tokens[1:])
	case cmd == "git" && len(tokens) > 2 && tokens[1] == "mv":
		files = firstFileArg(tokens[2:])
	case cmd == "unlink" && len(tokens) > 1 && !isShellSyntax(tokens[1]):
		files = tokens[1:2]
	}

	return resolvePaths(files, projectDir)
}

var readOnlyPrefixes = []string{
	"ls", "cat", "echo", "grep", "find", "head", "tail", "less", "more",
	"cd", "pwd", "which", "whereis", "type", "file", "stat", "wc",
	"git status", "git log", "git diff", "git show", "git branch",
	"git fetch", "git pull", "git push", "git clone", "git checkout",
	"git stash", "git remote", "git tag", "git rev-parse",
	"npm ", "yarn ", "pnpm ", "node ", "python", "pip ", "uv ",
	"cargo ", "go ", "make", "cmake", "docker ", "kubectl ",
	"curl ", "wget ", "ssh ", "scp ", "rsync ",
}

func isReadOnlyCommand(command string) bool {
	for _, prefix := range readOnlyPrefixes {
		if strings.HasPrefix(command, prefix) {
			return true
		}
	}
	return false
}

// fileArgs returns non-flag tokens up to the first shell operator or redirect.
func fileArgs(tokens []string) []string {
	var files []string
	for _, tok := range tokens {
		if isShellSyntax(tok) {
			break
		}
		if !strings.HasPrefix(tok, "-") {
			files = append(files, tok)
		}
	}
	return files
}

// firstFileArg returns the first non-flag token, stopping at shell syntax.
func firstFileArg(tokens []string) []string {
	if args := fileArgs(tokens); len(args) > 0 {
		return args[:1]
	}
	return nil
}

func resolvePaths(files []string, projectDir string) []string {
	var resolved []string
	for _, f := range files {
		if !filepath.IsAbs(f) {
			f = filepath.Join(projectDir, f)
		}
		resolved = append(resolved, filepath.Clean(f))
	}
	return resolved
}

func shellTokenize(s string) []string {
	s = strings.TrimSpace(s)
	if s == "" {
		return nil
	}

	var tokens []string
	var current strings.Builder
	inSingle := false
	inDouble := false

	for i := 0; i < len(s); i++ {
		ch := s[i]
		switch {
		case ch == '\'' && !inDouble:
			inSingle = !inSingle
		case ch == '"' && !inSingle:
			inDouble = !inDouble
		case ch == ' ' && !inSingle && !inDouble:
			if current.Len() > 0 {
				tokens = append(tokens, current.String())
				current.Reset()
			}
		default:
			current.WriteByte(ch)
		}
	}
	if current.Len() > 0 {
		tokens = append(tokens, current.String())
	}
	return tokens
}

var shellOperators = map[string]bool{
	"&&": true, "||": true, ";": true, "|": true,
}

var redirectPrefixes = []string{">", ">>", "<", "2>", "2>>", "1>", "1>>", "2>&1"}

func isShellSyntax(token string) bool {
	if shellOperators[token] {
		return true
	}
	for _, prefix := range redirectPrefixes {
		if strings.HasPrefix(token, prefix) {
			return true
		}
	}
	return false
}
